src/ogma/features/nodeGrouping.ts

```diff
@@ -0,0 +1,205 @@
+        groupIdFunction: (node) => {
+          if (this._isRuleNotApplicableToNode(node)) {
+            return undefined;
+          } else {
+            const propertyValue = node.getData(['properties', this.groupRule?.property ?? '']);
+            const originalValue =
+              typeof propertyValue === 'object'
+                ? (propertyValue as ConflictValue).original
+                : propertyValue;
+            // groupRule is defined if not we returned undefined
+            return `${this.groupRule!.type}-${originalValue}`.toLowerCase().trim();
+          }
+        },
+        nodeGenerator: (nodes) => {
+          return {
+            data: {
+              categories: [this.groupRule?.type],
+              properties: {
+                size: nodes.size
+              }
+            }
+          };
+        },
+        edgeGenerator: () => {
+          return {
+            data: {
+              type: LKE_NODE_GROUPING_EDGE
+            }
+          };
+        },
+        showContents: true,
+        duration: 300,
+        padding: 10
+      });
+      this._listenToTransformationEvents();
+    } else {
+      await this.refreshTransformation();
+    }
+  }
+
+  /**
+   * refresh the transformation
+   * Called when there is a change in the rule
+   */
+  public async refreshTransformation(): Promise<void> {
+    await this.transformation?.refresh();
+  }
+
+  /**
+   * init node grouping style
+   */
+  public initNodeGroupingStyle(): void {
+    this.nodeGroupingStyleRule = this._ogma.styles.addRule({
+      nodeAttributes: {
+        // Any default style will go here
+        text: {
+          content: (node: Node<LkNodeData> | undefined): string | undefined => {
+            return this._getNodeGroupingCaption(node);
+          },
+          style: 'bold'
+        },
+        layer: -1,
+        opacity: 0.32
+      },
+      nodeSelector: (node) => {
+        return node.isVirtual();
+      },
+      nodeDependencies: {self: {data: true}}
+    });
+  }
+
+  /**
+   * run layout on all subnodes of virtual nodes
+   */
+  public async runLayoutOnAllSubeNodes(): Promise<void> {
+    // @ts-ignore getContext exists on the transformation but hidden by the types
+    const virtualNodes = this.transformation.getContext().virtualNodes;
+    const rawNodesList = virtualNodes.getSubNodes();
+    const promisesList: Promise<void>[] = [];
+    for (let i = 0; i < rawNodesList.length; i++) {
+      const subNodes = rawNodesList[i];
+      if (subNodes !== undefined) {
+        promisesList.push(this._runSubNodesLayout(subNodes));
+      }
+    }
+    promisesList.push(this._runSubNodesLayout(virtualNodes));
+    await Promise.all(promisesList);
+  }
+
+  /**
+   * Run the layout on the subnodes of the virtual node
+   * @param subNodes nodes part of a virtual node
+   */
+  private async _runSubNodesLayout(subNodes: NodeList<LkNodeData, LkEdgeData>): Promise<void> {
```

Comment:
can we change this name, since it seems that you can run subNodes and virtual nodes with it? Maybe just runNodesLayout?